### PR TITLE
[Build/CI] introduce databend Dev container

### LINKED ISSUE #3845: dev container implementation

### PR DESCRIPTION
**Pull request by ZhiHanZ**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
1. support to run databend build && test with zero dependencies(only need to setup docker with proper permission write permission on /tmp and source code directory)
2. disable std io buffer during stateless tests(would introduce flaky issues on docker based environment, and should be compatible with currently setups)
3. support to cache intermediate results built from cargo, and use /tmp directory to get avoid of conflictions between native and docker based cargo environment

How to use

build binary artifacts
```bash
./scripts/setup/run_docker.sh  make build
```
run test
```bash
./scripts/setup/run_docker.sh  make test
```
debug or get into dev container
```bash
./scripts/setup/run_docker.sh /bin/bash
```
## Changelog

- New Feature
- Build/Testing/CI

## Related Issues

Fixes #3845 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by PsiACE** [comment]
macos :  FileNotFoundError: [Errno 2] No such file or directory: 'stdbuf'

File: scripts/setup/run_docker.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+SOURCE_DIR_MOUNT_DEST=/source
+# how does it work?
+# 1. docker build under root user with cargo installed on /opt/rust/cargo/bin
+# 2. to handle with permission issue on target release, we grant a user with same uid/gid to the host user for target dir access
+START_COMMAND=("/bin/bash" "-lc" "groupadd --gid $(id -g) -f databendgroup \
+  && useradd -o --uid $(id -u) --gid $(id -g) --no-create-home --home-dir /source databendbuild \
+  && chown -R databendbuild:databendgroup /source \
+  && chown -R databendbuild:databendgroup /opt/rust/cargo \
+  && chown -R databendbuild:databendgroup /tmp \
+  && sudo -EHs -u databendbuild bash -c 'cd /source && export LC_NUMERIC='de_DE.UTF-8' && export LC_ALL=C && DATABEND_DEV_CONTAINER=TRUE RUST_BACKTRACE=full env CARGO_HOME=/opt/rust/cargo PATH=/home/rust/.cargo/bin:/opt/rust/cargo/bin:/usr/local/musl/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin $*'")
```
Comment (Xuanwo) [inline scripts/setup/run_docker.sh:21]: How about use `export LC_NUMERIC='en_US.UTF-8' && export LC_ALL=en_US.UTF-8` here?

Keep the same with the value in container.

Comment (ZhiHanZ) [reply scripts/setup/run_docker.sh]: revised

File: docker/build-tool/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        python3-pip \
+        && \
+    apt-get clean && rm -rf /var/lib/apt/lists/* && \
+    useradd rust --user-group --create-home --shell /bin/bash --groups sudo
+# We use the instructions at https://github.com/rust-lang/rustup/issues/2383
+# to install the rustup toolchain as root.
+ENV RUSTUP_HOME=/opt/rust/rustup \
+    PATH=/home/rust/.cargo/bin:/opt/rust/cargo/bin:/usr/local/musl/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+
+RUN curl https://sh.rustup.rs -sSf -sSf |\
```
Comment (Xuanwo) [inline docker/build-tool/Dockerfile:33]:
```suggestion
RUN curl https://sh.rustup.rs -sSf |\
```

File: docker/build-tool/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    env CARGO_HOME=/opt/rust/cargo \
+        sh -s -- -y --default-toolchain stable --profile minimal --no-modify-path && \
+    export RUST_VERSION="$(awk -F'[ ="]+' '$1 == "channel" { print $2 }' /rust-toolchain.toml)" && \
+    env CARGO_HOME=/opt/rust/cargo rustup install $RUST_VERSION && \
+    env CARGO_HOME=/opt/rust/cargo rustup set profile minimal && \
+    env CARGO_HOME=/opt/rust/cargo rustup component add rustfmt --toolchain $RUST_VERSION && \
+    env CARGO_HOME=/opt/rust/cargo  rustup component add rust-src --toolchain $RUST_VERSION && \
```
Comment (Xuanwo) [inline docker/build-tool/Dockerfile:40]:
```suggestion
    env CARGO_HOME=/opt/rust/cargo rustup component add rust-src --toolchain $RUST_VERSION && \
```

How about keeping the same indent?

File: docker/build-tool/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ENV RUSTUP_HOME=/opt/rust/rustup \
+    PATH=/home/rust/.cargo/bin:/opt/rust/cargo/bin:/usr/local/musl/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+
+RUN curl https://sh.rustup.rs -sSf -sSf |\
+    env CARGO_HOME=/opt/rust/cargo \
+        sh -s -- -y --default-toolchain stable --profile minimal --no-modify-path && \
+    export RUST_VERSION="$(awk -F'[ ="]+' '$1 == "channel" { print $2 }' /rust-toolchain.toml)" && \
+    env CARGO_HOME=/opt/rust/cargo rustup install $RUST_VERSION && \
+    env CARGO_HOME=/opt/rust/cargo rustup set profile minimal && \
+    env CARGO_HOME=/opt/rust/cargo rustup component add rustfmt --toolchain $RUST_VERSION && \
```
Comment (Xuanwo) [inline docker/build-tool/Dockerfile:39]: Maybe we can use `rustup component add rustfmt rust-src clippy miri`?

File: docker/build-tool/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    env CARGO_HOME=/opt/rust/cargo  rustup component add rust-src --toolchain $RUST_VERSION && \
+    env CARGO_HOME=/opt/rust/cargo  rustup component add clippy --toolchain $RUST_VERSION && \
+    env CARGO_HOME=/opt/rust/cargo  rustup component add miri --toolchain $RUST_VERSION && \
+    env CARGO_HOME=/opt/rust/cargo rustup default "$RUST_VERSION"
+
+
+RUN mkdir -p /home/rust/libs /home/rust/src /home/rust/.cargo && \
+    ln -s /opt/rust/cargo/config /home/rust/.cargo/config
+
+RUN export DEBIAN_FRONTEND=noninteractive && apt-get update -y && apt-get install  -yq apt-utils && apt-get install -yq psmisc procps locales && python3 -m pip install boto3 "moto[all]" yapf shfmt-py coscmd PyYAML
```
Comment (Xuanwo) [inline docker/build-tool/Dockerfile:49]: Do we need to run `apt clean` here?

Comment (Xuanwo) [reply docker/build-tool/Dockerfile:49]: What does it mean for `apt-get install -yq`?

File: docker/build-tool/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,54 @@
+FROM debian:buster
+COPY ./dev_setup.sh /build_setup.sh
+COPY ./rust-toolchain.toml /rust-toolchain.toml
+RUN export DEBIAN_FRONTEND=noninteractive && apt-get update -y && \
```
Comment (Xuanwo) [inline docker/build-tool/Dockerfile:4]: Maybe it's better to set `ENV DEBIAN_FRONTEND=noninteractiv`?

File: docker/build-tool/Dockerfile (hunk 3 above)
Comment (ZhiHanZ) [reply docker/build-tool/Dockerfile:40]: looking for a Dockerfile linter on clion...

File: docker/build-tool/Dockerfile (hunk 6 above)
Comment (Xuanwo) [reply docker/build-tool/Dockerfile:4]: BTW, I believe the value should be `noninteractive`. @ZhiHanZ 

ref: https://www.debian.org/releases/sarge/s390/ch05s02.html.en

File: docker/build-tool/Dockerfile (hunk 5 above)
Comment (ZhiHanZ) [reply docker/build-tool/Dockerfile]: removed, locale came from surprising issues during stateless test and should be merged into a single RUN command

**Comment by flaneur2020** [comment]
すごい!

**Pull request merged**